arch/arm64/core/cortex_r/arm_mpu.c

```diff
@@ -183,4 +198,293 @@ void z_arm64_mm_init(bool is_primary_core)
+		}
+	}
+
+	return -1;
+}
+
+static int dup_dynamic_regions(struct dynamic_region_info *dst, int len)
+{
+	__ASSERT(sys_dyn_regions_num < len, "system dynamic region nums too large.");
+
+	int i;
+
+	for (i = 0; i < sys_dyn_regions_num; i++) {
+		dst[i] = sys_dyn_regions[i];
+	}
+	for (; i < len; i++) {
+		dst[i].index = -1;
+	}
+	return sys_dyn_regions_num;
 }
+
+static void set_region(struct arm_mpu_region *region,
+		       uint64_t base, uint64_t limit,
+		       struct arm_mpu_region_attr *attr)
+{
+	region->base = base;
+	region->limit = limit;
+	region->attr = *attr;
+}
+
+static int get_underlying_region_idx(struct dynamic_region_info *dyn_regions,
+				     uint8_t region_num, uint64_t base,
+				     uint64_t limit)
+{
+	for (uint8_t idx = 0; idx < region_num; idx++) {
+		struct arm_mpu_region *region = &(dyn_regions[idx].region_conf);
+
+		if (base >= region->base && limit <= region->limit) {
+			return idx;
+		}
+	}
+	return -1;
+}
+
+static uint8_t insert_region(struct dynamic_region_info *dyn_regions,
+			     uint8_t region_idx, uint8_t region_num,
+			     uintptr_t start, size_t size,
+			     struct arm_mpu_region_attr *attr)
+{
+	__ASSERT(region_idx < region_num,
+		 "Out-of-bounds error for dynamic region map. region idx: %d, region num: %d",
+		 region_idx, region_num);
+
+	/* base: inclusive, limit: exclusive */
+	uint64_t base = (uint64_t)start;
+	uint64_t limit = base + size;
+
+	int u_idx = get_underlying_region_idx(dyn_regions, region_idx, base, limit);
+
+	__ASSERT(u_idx >= 0, "Invalid underlying region index");
+
+	/* Get underlying region range and attr */
+	struct arm_mpu_region *u_region = &(dyn_regions[u_idx].region_conf);
+	uint64_t u_base = u_region->base;
+	uint64_t u_limit = u_region->limit;
+	struct arm_mpu_region_attr *u_attr = &u_region->attr;
+
+	/* Temporally holding new region available to be configured */
+	struct arm_mpu_region *curr_region = &(dyn_regions[region_idx].region_conf);
+
+	if (base == u_base && limit == u_limit) {
+		/*
+		 * The new region overlaps entirely with the
+		 * underlying region. Simply update the attr.
+		 */
+		set_region(u_region, base, limit, attr);
+	} else if (base == u_base) {
+		set_region(curr_region, base, limit, attr);
+		set_region(u_region, limit, u_limit, u_attr);
+		region_idx++;
+	} else if (limit == u_limit) {
+		set_region(u_region, u_base, base, u_attr);
+		set_region(curr_region, base, limit, attr);
+		region_idx++;
+	} else {
+		set_region(u_region, u_base, base, u_attr);
+		set_region(curr_region, base, limit, attr);
+		region_idx++;
+		curr_region = &(dyn_regions[region_idx].region_conf);
+		set_region(curr_region, limit, u_limit, u_attr);
+		region_idx++;
+	}
+
+	return region_idx;
+}
+
+static int flush_dynamic_regions_to_mpu(struct dynamic_region_info *dyn_regions,
+					uint8_t region_num)
+{
+	int reg_avail_idx = static_regions_num;
+	/*
+	 * Clean the dynamic regions
+	 */
+	for (int i = reg_avail_idx; i < get_num_regions(); i++) {
```

Comment:
nit: pick a consistent type for the `for` cycles. You used `int` here but `uint8_t` previously. Also decide if you want to declare the variable inside the `for` or outside. In general my suggestion is always to use `size_t`.

Reply:
Ok, I'll change it to size_t